skt/__init__.py

```diff
@@ -137,41 +137,62 @@ def __init__(self, uri, ref=None, wdir=None):
             wdir:   The directory to house the clone and to checkout into.
                     Creates and uses a temporary directory if not specified.
         """
-        # FIXME Move expansion up the call stack, as this limits the class
-        # usefulness, because tilde is a valid path character.
-        # The git "working directory" (the "checkout")
-        self.wdir = (os.path.expanduser(wdir)
-                     if wdir is not None
-                     else tempfile.mkdtemp())
-        # The cloned git repository
-        self.gdir = "%s/.git" % self.wdir
+        # Get the full path to the work directory and .git directory.
+        self.wdir = get_workdir(wdir)
+        self.gdir = self.get_gitdir(wdir)
         # The origin remote's URL
         self.uri = uri
         # The remote reference to checkout
         self.ref = ref if ref is not None else "master"
         self.info = []
         self.mergelog = "%s/merge.log" % self.wdir
 
-        try:
-            os.mkdir(self.wdir)
-        except OSError:
-            pass
+        # Create the work directory
+        self.create_workdir()
 
-        try:
-            os.unlink(self.mergelog)
-        except OSError:
-            pass
+        # Remove the old merge log
+        self.remove_mergelog()
+
+        # Set the origin remote for the repository
+        self.set_repo_remote()
+
+        logging.info("base repo url: %s", self.uri)
+        logging.info("base ref: %s", self.ref)
+        logging.info("work dir: %s", self.wdir)
 
-        self.git_cmd("init")
+    def initialize_repo():
+        """Initialize the git repository"""
+        git_directory = "{}/.git".format(self.wdir)
+        if not os.path.isdir(git_directory):
+            self.git_cmd("init")
 
+    def set_repo_remote():
+        """Set the origin remote for the repository"""
         try:
             self.git_cmd("remote", "set-url", "origin", self.uri)
         except subprocess.CalledProcessError:
             self.git_cmd("remote", "add", "origin", self.uri)
 
-        logging.info("base repo url: %s", self.uri)
-        logging.info("base ref: %s", self.ref)
-        logging.info("work dir: %s", self.wdir)
+    def get_workdir(workdir):
+        """Determine the full path to the work directory"""
+        if workdir:
+            return os.path.expanduser(workdir)
+        else:
+            return tempfile.mkdtemp()
+
+    def get_gitdir(workdir):
+        """Get the path to the .git directory within the work directory"""
+        return "%s/.git" % workdir
+
+    def create_workdir():
+        """Create the work directory if it does not exist"""
+        if not os.path.is_dir(self.wdir):
+            os.mkdir(self.wdir)
```

Comment:
I'm not sure we need a public function for creating a workdir. I doubt the class user would want to call it.

Also, please just run `os.mkdir` and then handle the "already exists" exception, instead of checking, then creating. This way we would avoid a race condition.